source/pdal/pdalc_pipeline.cpp

```diff
@@ -26,70 +26,73 @@
  * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
  * POSSIBILITY OF SUCH DAMAGE.
  *****************************************************************************/
-
+#define _CRT_SECURE_NO_WARNINGS
 #include "pdalc_pipeline.h"
 
 #include <string>
 
-#include <pdal/PipelineExecutor.hpp>
-
 #include "pdalc_pointviewiterator.h"
 
-// TODO Address cause of std::min problems. See https://github.com/PDAL/CAPI/issues/4
+#include <pdal/util/Utils.hpp>
+#include <pdal/PipelineWriter.hpp>
+#include <pdal/Stage.hpp>
+#include <pdal/pdal_types.hpp>
+
 #undef min
 
 namespace pdal
 {
 namespace capi
 {
+
 extern "C"
 {
-    PDALPipelinePtr PDALCreatePipeline(const char* json)
+    struct Pipeline
     {
-        PDALPipelinePtr pipeline = nullptr;
+    public:
+
+        PipelineManagerPtr manager = std::make_unique<PipelineManager>();
 
+        bool m_executed = false;
+        std::stringstream logStream;
+    };
+
+    PDALPipelinePtr PDALCreatePipeline(const char* json)
+    {
+        Pipeline* pipeline = new Pipeline();
         if (json && std::strlen(json) > 0)
         {
-            pdal::PipelineExecutor *executor = nullptr;
-
             try
             {
-                pdal::PipelineExecutor stackpipe(json);
-                executor = new pdal::PipelineExecutor(json);
+                std::stringstream* s = &pipeline->logStream;
+                LogPtr lptr(pdal::Log::makeLog("pdal capi", s, true));
+                pipeline->manager->setLog(lptr);
+
+                std::stringstream strm;
+                strm << json;
+                pipeline->manager->readPipeline(strm);
             }
             catch (const std::exception &e)
             {
                 printf("Could not create pipeline: %s\n%s\n", e.what(), json);
-                executor = nullptr;
+                delete pipeline;
+                return nullptr;
             }
 
-            if (executor)
+            try
             {
-                bool valid = false;
-
-                try
-                {
-                    valid = executor->validate();
-                }
-                catch (const std::exception &e)
-                {
-                    printf("Error while validating pipeline: %s\n%s\n", e.what(), json);
-                }
-
-                if (valid)
-                {
-                    pipeline = new Pipeline(executor);
-                }
-                else
-                {
-                    delete executor;
-                    executor = NULL;
-                    printf("The pipeline is invalid:\n%s\n", json);
-                }
+                pipeline->manager->prepare();
+            }
+            catch (const std::exception &e)
+            {
+                printf("Error while validating pipeline: %s\n%s\n", e.what(), json);
+                delete pipeline;
```

Comment:
unique_ptr would mean you wouldn't have to remember to clean these up.

Reply:
This is WAI.